internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,260 @@
+				rateLimitInfo.Remaining, rateLimitInfo.Limit,
+				float64(rateLimitInfo.Remaining)/float64(rateLimitInfo.Limit), m.threshold))
+		}
+	}
+
+	m.mu.Lock()
+	m.lastRequestTime = time.Now()
+	m.mu.Unlock()
+
+	m.log("DEBUG: RequestMiddleware completed successfully\n")
+
+	return nil
+}
+
+func (m *Manager) ResponseMiddleware(client *resty.Client, resp *resty.Response) error {
+	m.log(fmt.Sprintf("DEBUG: ResponseMiddleware called, enabled: %v\n", m.enabled))
+
+	if !m.enabled {
+		m.log("DEBUG: Rate limiting disabled - response middleware returning early\n")
+		return nil
+	}
+
+	m.log("DEBUG: Setting up defer function for semaphore release\n")
+
+	defer func() {
+		m.log("DEBUG: Defer function executing - attempting to release semaphore\n")
+
+		select {
+		case <-m.requestSemaphore:
+			m.log("DEBUG: Successfully released semaphore slot\n")
+		case <-time.After(1 * time.Second):
+			m.log("DEBUG: Semaphore release timeout - continuing anyway\n")
+		}
+
+		m.log("DEBUG: Updating active request count\n")
+
+		m.mu.Lock()
+		m.activeRequests--
+		if m.activeRequests < 0 {
+			m.activeRequests = 0 // Prevent negative count
+		}
+		m.log(fmt.Sprintf("DEBUG: Active requests now: %d\n", m.activeRequests))
+		m.mu.Unlock()
+
+		m.log("DEBUG: Defer function completed\n")
+	}()
+
+	m.log("DEBUG: Extracting rate limit headers\n")
+
+	limitHeader := resp.Header().Get("x-ratelimit-limit")
+	periodHeader := resp.Header().Get("x-ratelimit-period")
+	remainingHeader := resp.Header().Get("x-ratelimit-remaining")
+	resetHeader := resp.Header().Get("x-ratelimit-reset")
+
+	m.log(fmt.Sprintf("DEBUG: Rate limit headers - limit: %q, period: %q, remaining: %q, reset: %q\n",
+		limitHeader, periodHeader, remainingHeader, resetHeader))
+
+	if limitHeader != "" && remainingHeader != "" {
+		m.log("DEBUG: Parsing rate limit headers\n")
+
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		rateLimitInfo := &RateLimitInfo{}
+
+		if limit, err := strconv.Atoi(limitHeader); err == nil {
+			rateLimitInfo.Limit = limit
+		}
+		if period, err := strconv.Atoi(periodHeader); err == nil {
+			rateLimitInfo.Period = period
+		}
+		if remaining, err := strconv.Atoi(remainingHeader); err == nil {
+			rateLimitInfo.Remaining = remaining
+		}
+		if reset, err := strconv.Atoi(resetHeader); err == nil {
+			rateLimitInfo.Reset = reset
+		}
+
+		m.rateLimitInfo = rateLimitInfo
+
+		m.log(fmt.Sprintf("DEBUG: Parsed rate limit info: %+v\n", rateLimitInfo))
+	} else {
+		m.log("DEBUG: No rate limit headers found or incomplete\n")
+	}
+
+	m.log("DEBUG: ResponseMiddleware completed successfully\n")
+
+	return nil
+}
+
+func (m *Manager) calculateDelay(rateLimitInfo *RateLimitInfo) time.Duration {
+	if rateLimitInfo.Remaining <= 0 && rateLimitInfo.Reset > 0 {
+		delay := time.Duration(rateLimitInfo.Reset) * time.Second
+		if delay > 2*time.Minute {
+			delay = 2 * time.Minute
+		}
+		jitter := time.Duration(float64(delay) * 0.1)
```

Comment:
Shouldn't `jitter` be random?

Reply:
Also, you can just have `delay` always be a `float64` and do `return time.Duration(delay * 1.1)`

Reply:
Fixed!